src/main/java/io/confluent/connect/jdbc/sink/metadata/FieldsMetadata.java

```diff
@@ -161,19 +160,29 @@ public static FieldsMetadata extract(
           }
           keyFieldNames.addAll(configuredPkFields);
         }
+        for (String fieldName : keyFieldNames) {
+          final Schema fieldSchema = valueSchema.field(fieldName).schema();
+          allFields.put(fieldName, new SinkRecordField(fieldSchema.type(), fieldName, true, false, fieldSchema.defaultValue()));
+        }
       }
       break;
 
     }
 
+    final Set<String> nonKeyFieldNames = new LinkedHashSet<>();
     if (valueSchema != null) {
-      for (Field field : valueSchema.fields()) {
-        final boolean isKeyField = keyFieldNames.contains(field.name());
-        if (!isKeyField) {
-          nonKeyFieldNames.add(field.name());
+      for (Field field: valueSchema.fields()) {
+        if (keyFieldNames.contains(field.name())) {
+          continue;
+        }
+        if (!fieldsWhitelist.isEmpty() && !fieldsWhitelist.contains(field.name())) {
+          continue;
         }
+
+        nonKeyFieldNames.add(field.name());
+
         final Schema fieldSchema = field.schema();
-        allFields.put(field.name(), new SinkRecordField(fieldSchema.type(), field.name(), isKeyField, !isKeyField && fieldSchema.isOptional(), fieldSchema.defaultValue()));
+        allFields.put(field.name(), new SinkRecordField(fieldSchema.type(), field.name(), false, fieldSchema.isOptional(), fieldSchema.defaultValue()));
```

Comment:
It seems like the behavior was changed here? `allFields` doesn't actually contain all the fields anymore (which is pretty clear since one of the flags is always `false`). Either the meaning of `allFields` changed and we should probably rename to clarify, or we may not have test coverage for something (it looks like it is used, for example, to determine if we need to alter the table).